src/env/root/mod.rs

```diff
@@ -59,9 +59,28 @@ impl<'a> RootRuntime<'a> {
             call_targets: Default::default(),
             call_stack: Default::default(),
             buffer: Default::default(),
+
+            #[cfg(feature = "debug")]
+            logger: RefCell::new(Box::new(|s: &str| {
+                println!("{}", s);
+            })),
         }))
     }
 
+    #[cfg(feature = "debug")]
+    pub fn set_logger(&mut self, f: Box<dyn Fn(&str)>) {
+        let mut logger = self.0.logger.borrow_mut();
+        *logger = f;
+    }
+
+    #[cfg(feature = "debug")]
+    pub(crate) fn print(&self, bytes: &[u8]) {
+        match String::from_utf8(bytes.to_vec()) {
```

Comment:
Why not [`from_utf8_lossy`](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8_lossy)?

Reply:
Good call, I should just use that.